### PR TITLE
MRG Explicitly use UTF8 for sidecar I/O

### LINKED ISSUE #622: UnicodeEncodeError: 'charmap' codec can't encode character '\xf6' in position 147
Hello, everyone!

I was moving from 0.4 to 0.5 version of mne-bids and faced with a problem while writing files in BIDS format.

The code is next:

```
import mne_bids
import mne
from pathlib import Path

proj_root = Path() / '..'
data_raw_dir = proj_root / 'data_raw'
data_bids_dir = proj_root / 'data_bids'

raw_file = mne.io.read_raw_fif(str(raw_file_path)) # raw_file_path is just a direct path to MEG fif file

subj_id = '1001'
date_record = '20201111'
task_name = 'vid1'

bids_filename = mne_bids.BIDSPath(subject=subj_id, session=date_record, task=task_name, root=data_bids_dir)
mne_bids.write_raw_bids(raw=raw_file, bids_path=bids_filename, verbose=True,  anonymize={'daysback': 40000}, overwrite=True)

```

Output error:

```
Traceback (most recent call last):
  File "<input>", line 8, in <module>
  File "E:\Egor_Levchenko\Software\Miniconda\envs\MCCA-MEG\lib\site-packages\mne_bids\write.py", line 1082, in write_raw_bids
    _readme(bids_path.datatype, readme_fname, False, verbose)
  File "E:\Egor_Levchenko\Software\Miniconda\envs\MCCA-MEG\lib\site-packages\mne_bids\write.py", line 236, in _readme
    _write_text(fname, text, overwrite=True, verbose=verbose)
  File "E:\Egor_Levchenko\Software\Miniconda\envs\MCCA-MEG\lib\site-packages\mne_bids\utils.py", line 176, in _write_text
    fid.write(text)
  File "E:\Egor_Levchenko\Software\Miniconda\envs\MCCA-MEG\lib\encodings\cp1251.py", line 19, in encode
    return codecs.charmap_encode(input,self.errors,encoding_table)[0]
UnicodeEncodeError: 'charmap' codec can't encode character '\xf6' in position 147: character maps to <undefined>
C:\Programs\JetBrains\PyCharm Community Edition 2020.1\plugins\python-ce\helpers\pydev\_pydevd_bundle\pydevd_thrift.py:322: RuntimeWarning: Could not locate a data file of a supported format. This is likely a problem with your BIDS dataset. Please run the BIDS validator on your data. (root=..\data_bids, basename=sub-1001_ses-20201111_task-vid1). []
  value = format % v
```

I am building my environment using conda with this `.yml` file:
```
name: MCCA-MEG
dependencies:
  - python=3.7.7
  - pip=20.2.4
  - numpy=1.17.2
  - scipy=1.3.1
  - matplotlib=3.3.2
  - pandas=1.1.1
  - conda-forge::mne=0.21.0
  - conda-forge::mne-bids=0.5
  - conda-forge::dicom2nifti=2.2.10 #should be deleted if dcm2niix is successful enough
  - conda-forge::dcm2niix=1.0.20190902
  - pip:
      - pybids==0.12.3
```

Can you tell me please what am I missing in the new version of mne-bids?

Thanks for any suggestions!

### PR DESCRIPTION
**Pull request by hoechenberger**
PR Description
--------------

For reading, use `utf-8-sig` to take care of a BOM if present (even though it's illegal in JSON – this makes our readers more
robust).

For writing, always use `utf-8`, following BIDS (btw BIDS is not explicit about the BOM – just says "UTF8").

Closes  #622

Merge checklist
---------------

Maintainer, please confirm the following before merging:

- [ ] All comments resolved
- [ ] This is not your own PR
- [ ] All CIs are happy
- [ ] PR title starts with [MRG]
- [ ] [whats_new.rst](https://github.com/mne-tools/mne-bids/blob/master/doc/whats_new.rst) is updated
- [ ] PR description includes phrase "closes <#issue-number>"

### CODE REVIEW

**Review by sappelhoff — COMMENTED** [review]
I like the idea and the PR, just two questions:

why do you sometimes use `utf-8-sig`? What's that, and why is it not always used?

what is a BOM (byte order mark) in the UTF-8 context, and what should BIDS specify that it currently doesn't? :)

**Comment by hoechenberger** [comment]
`utf-8-sig` is the charset in Python that 

- when writing, adds the BOM (it's just a byte at the beginning of the file that is supposed to indicate that "what follows is unicode", and is rendered as a zero-space character)
- when reading and encountering a BOM, it strips the BOM (as one is supposed to); and if there's no BOM, it doesn't strip anything. So this makes reading robust as it works with both BOM and non-BOM files.

BIDS just specifies "UTF8", which to me is clear means: no BOM; but for example Excel would always add a BOM when saving a UTF8 text file. (Note that **reading** a UTF8 file without BOM in Excel makes Excel think it's ANSI… But that's really something MS ought to fix.)

JSON files must not contain a BOM, according to the standard which I'm now too lazy to fetch ;)

So:
- to be robust, we should use `utf-8-sig` to correctly handle the BOM if there is one
- for writing, we should not add a BOM: it's really a superfluous annoying thing
- BIDS should probably be amended to explicate what "UTF8" means – with (*shiver*) or without (yay!) BOM

**Comment by hoechenberger** [comment]
What we could also do is use `utf-8-sig` for writing too, to make life easier for Windows users… PsychoPy uses that charset too (UTF8 with BOM)

**Comment by hoechenberger** [comment]
> What we could also do is use `utf-8-sig` for writing too, to make life easier for Windows users… PsychoPy uses that charset too (UTF8 with BOM)

See https://stackoverflow.com/a/11784679

**Comment by hoechenberger** [comment]
> What we could also do is use `utf-8-sig` for writing too,

Only for the non-JSON sidecars, as the BOM is illegal in JSON:
https://stackoverflow.com/a/38036753

**Review by agramfort — APPROVED** [review]
nice !

**Comment by agramfort** [comment]
@sappelhoff merge if happy

File: examples/convert_eeg_to_bids.py [hunk 1]
```diff
@@ -206,7 +206,7 @@
 # If you are preparing a manuscript, please make sure to also cite MNE-BIDS
 # there.
 readme = os.path.join(bids_root, 'README')
-with open(readme, 'r') as fid:
+with open(readme, 'r', encoding='utf-8') as fid:
```
Comment (sappelhoff) [inline examples/convert_eeg_to_bids.py:209]: why is this not `utf-8-sig`?

**Review by sappelhoff — COMMENTED** [review]
Okay, thanks a lot for the explanation!

But if `utf-8-sig` is the "robust thing" to do during reading, why don't we use it everywhere?

**Comment by sappelhoff** [comment]
> BIDS should probably be amended to explicate what "UTF8" means – with (shiver) or without (yay!) BOM

I would say let's wait with this until a person raises an issue.

> What we could also do is use utf-8-sig for writing too, to make life easier for Windows users… PsychoPy uses that charset too (UTF8 with BOM) for this reason

I think I would not do that right now ... or would #622 have been prevented through that?

**Comment by sappelhoff** [comment]
Wikipedia tells me that "for UTF-8, BOM is not necessary, but allowed" ... that's an argument for me to not include it :thinking:

**Comment by hoechenberger** [comment]
> Wikipedia tells me that ["for UTF-8, BOM is not necessary, but allowed"](https://en.wikipedia.org/wiki/Byte_order_mark#UTF-8) ... that's an argument for me to not include it 🤔

Yes so the IMHO technically cleanest solution is to NOT write the BOM.

But this will make it more difficult for Windows users to open the sidecar files e.g. in Excel or Notepad or whatever… do we care about that?

**Comment by sappelhoff** [comment]
I guess that's a fair point. I'm okay if we write the BOM for all non-json files.

File: examples/convert_eeg_to_bids.py (hunk 1 above)
Comment (sappelhoff) [reply examples/convert_eeg_to_bids.py:209]: Also, does reading with utf-8 ( no sig ) fail if a BOM is there?

**Comment by jasmainak** [comment]
can we add a test to avoid a regression in the future?

**Comment by hoechenberger** [comment]
> can we add a test to avoid a regression in the future?

I will see what I can do :)

**Comment by agramfort** [comment]
@sappelhoff merge if happy

**Review by sappelhoff — APPROVED** [review]

**Comment by sappelhoff** [comment]
Thanks @hoechenberger! Love the Döner Test.

**Pull request merged**